Fix shape_inference for V-schedules

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147000

I was hitting a hang in shape_inference when testing v-shaped schedules with >2 ranks in titan.

`self.next_rank` and `self.prev_rank` are used in shape inference but are not accurate for v-shaped schedules:
https://github.com/pytorch/pytorch/blob/bfcce6984b033640e01a647c44a8a13f86d64f5a/torch/distributed/pipelining/stage.py#L1325-L1326

Will clean up / delete the use of next_rank / prev rank in follow up PRs


cc @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

H-Huang committed:
Fix shape_inference for ZBV

[ghstack-poisoned]

wconstab commented on torch/distributed/pipelining/stage.py:
```diff
@@ -111,6 +111,10 @@ def _make_tensor_from_meta(
     )
 
 
+def _rank_to_global_rank(peer_rank: int, group: Optional[dist.ProcessGroup]):
```
i think you can just directly use this helper from distributed_c10d.py?

```
def get_global_rank(group: ProcessGroup, group_rank: int) -> int:
    """
    Translate a group rank into a global rank.

    ``group_rank`` must be part of `group` otherwise this raises RuntimeError.

    Args:
        group (ProcessGroup): ProcessGroup to find the global rank from.
        group_rank (int): Group rank to query.

    Returns:
        Global rank of ``group_rank`` relative to ``group``

    N.B. calling this function on the default process group returns identity
    """
```

H-Huang replied:
oops yeah, good point!

wconstab commented on torch/distributed/pipelining/stage.py:
```diff
@@ -1371,7 +1375,12 @@ def _shape_inference(
-                objects, src=self.prev_rank, group=self.group, device=self.device
+                objects,
+                src=_rank_to_global_rank(
+                    self.stage_index_to_group_rank[self.stage_index - 1], self.group
```
you mentioned another PR to remove self.next_rank / prev_rank.  I think another option is to convert those to @property functions, and implement them with something like this code snippet.

```self.stage_index_to_group_rank[self.stage_index - 1]```

We should also audit whether we always return group-rank or global-rank from that helper and make sure that's consistent or theres a flag for controlling it if we need it.

H-Huang committed:
Update on "Fix shape_inference for V-schedules"


I was hitting a hang in shape_inference when testing v-shaped schedules with >2 ranks in titan.

`self.next_rank` and `self.prev_rank` are used in shape inference but are not accurate for v-shaped schedules:
https://github.com/pytorch/pytorch/blob/bfcce6984b033640e01a647c44a8a13f86d64f5a/torch/distributed/pipelining/stage.py#L1325-L1326

Will clean up / delete the use of next_rank / prev rank in follow up PRs


cc awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge
